### ISSUE TITLE
Nested mock breaks when called as function in specs2 test

### ISSUE THREAD

**Issue by danielschleindlsperger** [mockito/mockito-scala#227]
Hello!

I'm using the Play! framework with specs2 was running into a problem after updating Play! (2.6 -> 2.8), sbt (1.2.8 -> 1.3.8) and Scala (2.12.7 -> 1.13.1).

The Problem is that mocks that should return another mock that was created in a function now longer work when that function was called inline.

This will fail:
```scala
def mockBar(): Bar = {
  val bar = mock[Bar]
  // ...
  bar
}

val foo = mock[Foo]
foo.bar returns mockBar()
```

While this will work:
```scala
def mockBar(): Bar = { /* same as above */}

val foo = mock[Foo]
val bar = mockBar()
foo.bar returns bar
```

Complete example spec:

```scala
package foo

import org.specs2.mock.Mockito
import org.specs2.mutable.Specification

case class Foo() {
  def bar(): String = "bar"
}

case class Baz(foo: Foo)

class FooSpec extends Specification with Mockito {

  def mockedFoo(): Foo = {
    val foo = mock[Foo]
    foo.bar returns "mockedBar"
    foo
  }

  "given a foo" >> {
    "returns bar with inline mock (broken)" >> {
      val baz = mock[Baz]
      baz.foo returns mockedFoo() // inline call
      val value = baz.foo.bar()
      value mustEqual "mockedBar"
    }

    "returns bar with intermediate val (works)" >> {
      val baz = mock[Baz]
      val mocked = mockedFoo()
      baz.foo returns mocked // intermediate val
      val value = baz.foo.bar()
      value mustEqual "mockedBar"
    }
  }
}
```

It fails with a mockito `MissingMethodInvocationException`:
```
[error] An exception was raised during the creation of the specification:
[error] when() requires an argument which has to be 'a method call on a mock'.
[error] For example:
[error]     when(mock.getArticles()).thenReturn(articles);
[error]
[error] Also, this error might show up because:
[error] 1. you stub either of: final/private/equals()/hashCode() methods.
[error]    Those methods *cannot* be stubbed/verified.
[error]    Mocking methods declared on non-public parent classes is not supported.
[error] 2. inside when() you don't call method on mock but on some other object.
[error] .
[error] This means that you have some code which should be enclosed in an example. Instead of writing:
[error]
[error]  "this is a block of examples" in {
[error]    // set-up something
[error]    createDatabase
[error]    "first example" in { 1 must_== 1 }
[error]    "second example" in { 1 must_== 1 }
[error]  }
[error]
[error] You should write:
[error]
[error]  "this is a block of examples" in {
[error]    "the setup must be ok" in {
[error]      createDatabase must not(throwAn[Exception])
[error]    }
[error]    "first example" in { 1 must_== 1 }
[error]    "second example" in { 1 must_== 1 }
[error]  }
[error]
[error]  Be careful because in the specification above the expectation might be that the
[error]  database will be created before the "first" and "second" examples. This will NOT be the case
[error]  unless you mark the specification as `sequential`. You can also have a look at the `BeforeEach/BeforeAll` traits
[error]  to implement this kind of functionality.
[error]
[error] EXCEPTION
[error]
[error] ============================================================
[error] org.mockito.exceptions.misusing.MissingMethodInvocationException:
[error] when() requires an argument which has to be 'a method call on a mock'.
[error] For example:
[error]     when(mock.getArticles()).thenReturn(articles);
[error]
[error] Also, this error might show up because:
[error] 1. you stub either of: final/private/equals()/hashCode() methods.
[error]    Those methods *cannot* be stubbed/verified.
[error]    Mocking methods declared on non-public parent classes is not supported.
[error] 2. inside when() you don't call method on mock but on some other object.
[error]
[error] ------------------------------------------------------------
[error]     org.mockito.exceptions.misusing.MissingMethodInvocationException:
[error]     when() requires an argument which has to be 'a method call on a mock'.
[error]     For example:
[error]         when(mock.getArticles()).thenReturn(articles);
[error]
[error]     Also, this error might show up because:
[error]     1. you stub either of: final/private/equals()/hashCode() methods.
[error]        Those methods *cannot* be stubbed/verified.
[error]        Mocking methods declared on non-public parent classes is not supported.
[error]     2. inside when() you don't call method on mock but on some other object.
```

Am I doing something wrong or this is a new limitation of the library? 😅

**Comment by ultrasecreth** [comment]
Hi @danielschleindlsperger, we receently made some changes on the macros that may have introduced this issue, I'll take a look to understand why is this happening and hopefully get a fix.

**Comment by ultrasecreth** [comment]
Hi @danielschleindlsperger, I just found a couple of things

a) you don't seem to be using mockito-scala by the look of your imports ;)
b) what you want to do would work, but only if the function that creates the mock is not also stubbing it, as you can't stub 2 mocks at the same time. 

I think that to solve your problem, and even make your tests simpler, you could try DeepStubs, for example, your test would become:

```scala
import org.mockito.specs2.Mockito
import org.mockito.DefaultAnswers
import org.specs2.mutable.Specification

case class Foo() {
  def bar(): String = "bar"
}

case class Baz(foo: Foo)

class FooSpec extends Specification with Mockito {
  "given a foo" >> {
    "returns bar with deep stub" >> {
      val baz = mock[Baz](DefaultAnswers.ReturnsDeepStubs)

      baz.foo.bar returns "mockedBar"

      baz.foo.bar mustEqual "mockedBar"
    }
  }
}
```

As you can see, mockito will create the intermediate mocks to satisfy any method call where the return type can be mocked, and that also works for stubbing it, so you can just express the call chain that your app will execute.

Note that in this case, I'm using mockito-scala and not specs2-mock (the import of the Mockito trait is `org.mockito.specs2.Mockito` instead of `org.specs2.mock.Mockito`)

Please let me know if that solves your problem

**Comment by danielschleindlsperger** [comment]
Hey @bbonanno 

You were right, apparently I'm not using mockito-scala, I'm sorry! 😅 I find the Play! framework and specs2 make it a little opaque which mockito library you actually end up using.

I've resorted to rewriting the tests to not have two nested mocks. That was a little time-consuming but I guess better than including a different version of mockito and rewriting everything.

I tried your suggestion with the "deep mocks" though and might use it in the future. Thank you!
